src/main/java/edu/ucla/library/iiif/auth/handlers/AccessCookieHandler.java

```diff
@@ -0,0 +1,130 @@
+import edu.ucla.library.iiif.auth.services.DatabaseService;
+
+import info.freelibrary.util.HTTP;
+
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+import io.vertx.core.http.Cookie;
+import io.vertx.core.http.HttpServerRequest;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.web.RoutingContext;
+import io.vertx.ext.web.templ.handlebars.HandlebarsTemplateEngine;
+
+/**
+ * Handler that handles access cookie requests.
+ */
+public class AccessCookieHandler implements Handler<RoutingContext> {
+
+    /**
+     * The application configuration.
+     */
+    private final JsonObject myConfig;
+
+    /**
+     * The service proxy for accessing the database.
+     */
+    private final DatabaseService myDatabaseServiceProxy;
+
+    /**
+     * The template engine for rendering the response.
+     */
+    private final HandlebarsTemplateEngine myHtmlTemplateEngine;
+
+    /**
+     * A {@link Trie} of Campus Network subnets.
+     */
+    private final Cidr4Trie<String> myCampusNetworkSubnets;
+
+    /**
+     * A service for generating and decrypting encrypted access cookies.
+     */
+    private final AccessCookieCryptoService myAccessCookieCryptoService;
+
+    /**
+     * Creates a handler that retrieves an access cookie for the client.
+     *
+     * @param aVertx The Vert.x instance
+     * @param aConfig A configuration
+     */
+    public AccessCookieHandler(final Vertx aVertx, final JsonObject aConfig) {
+        myConfig = aConfig;
+        myDatabaseServiceProxy = DatabaseService.createProxy(aVertx);
+        myHtmlTemplateEngine = HandlebarsTemplateEngine.create(aVertx);
+        myCampusNetworkSubnets = new Cidr4Trie<>();
+        myAccessCookieCryptoService = AccessCookieCryptoService.createProxy(aVertx);
+
+        for (final String subnet : aConfig.getString(Config.CAMPUS_NETWORK_SUBNETS).split(COMMA)) {
+            final Cidr4 cidr = new Cidr4(subnet);
+
+            // The value of the key doesn't particularly matter, since the methods we call on the Trie don't use it
+            myCampusNetworkSubnets.put(cidr, cidr.getAddressRange());
+        }
+    }
+
+    @Override
+    public void handle(final RoutingContext aContext) {
+        final HttpServerRequest request = aContext.request();
+
+        try {
+            final Ip4 clientIpAddress = new Ip4(request.remoteAddress().hostAddress());
+            final URI origin = URI.create(request.getParam(Param.ORIGIN));
+            final boolean isOnCampusNetwork = isOnNetwork(clientIpAddress, myCampusNetworkSubnets);
+
+            myDatabaseServiceProxy.getDegradedAllowed(origin.toString()).compose(isDegradedAllowed -> {
+                final Future<String> cookieGeneration = myAccessCookieCryptoService
+                        .generateCookie(clientIpAddress.getAddress(), isOnCampusNetwork, isDegradedAllowed);
+
+                return cookieGeneration.compose(cookieValue -> {
```

Comment:
Not a problem, but I need a quick tutorial... how does a `return` inside a lambda expression inside a `return void` method work?

Reply:
This part of the Vert.x docs has an overview: https://vertx.io/docs/vertx-core/java/#_future_composition

In this case, three Futures are being chained sequentially:
1. Querying the database
2. Generating a cookie
3. Rendering the HTML template

If all of these complete, the function passed to `onSuccess` (L108) is called; otherwise (i.e., if any of them fail), the function passed to `onFailure` (L112) is called.

Anytime a mapper function is passed to `compose( ... )`, it will return a Future.

(Strictly speaking, steps 2 and 3 don't _have_ to be sequential (i.e., their order doesn't really matter, so they can be composed with a CompositeFuture instead), but I figured it was clearer to write it the way I did. Although, I can change this if desired.)